src/Chainweb/Pact/PactService.hs

```diff
@@ -418,40 +422,80 @@ finalizeCheckpointer finalize = do
+    | otherwise = do
+        bs <- balances txs
+        let gt = gasTotals bs txs
+        V.mapM (validate bs gt) txs
   where
-    checkAccount validators tx = do
-      let !pm = P._pMeta $ payloadObj $ P._cmdPayload tx
-      let sender = P._pmSender pm
-          (P.GasLimit (P.ParsedInteger limit)) = P._pmGasLimit pm
-          (P.GasPrice (P.ParsedDecimal price)) = P._pmGasPrice pm
-          limitInCoin = price * fromIntegral limit
-      m <- readCoinAccount dbEnv sender
-      case m of
-        Nothing -> return True
-        Just (T2 b _g) -> do
-          let validations = (const (b >= limitInCoin)) : validators
-          return $! all ($ tx) validations
-
-    -- skip validation for genesis -- gas accounts / etc don't exist yet
-    checkOne tx = if bh == 0 then return True else checkOneReal tx
-    checkOneReal tx = do
-        let pactHash = view P.cmdHash tx
-        mb <- _cpLookupProcessedTx cp pactHash
-        if mb == Nothing
-            -- prop_tx_ttl_newblock_validate
-            then checkAccount [checkTimes] tx
-            else return False
+    validate :: Balances -> GasTotals -> ChainwebTransaction -> IO Bool
+    validate bs gt tx = case (,) <$> HM.lookup sender bs <*> HM.lookup sender gt of
+        Nothing -> pure False
+        Just (b, g) -> do
+            let validations = [gasCheck b g, checkTimes]
+            return $! all ($ tx) validations
+      where
+        gasCheck :: Decimal -> Decimal -> b -> Bool
+        gasCheck b g _ = b >= g
+
+        sender = P._pmSender . P._pMeta . payloadObj $ P._cmdPayload tx
+
+    checkTimes :: ChainwebTransaction -> Bool
+    checkTimes = timingsCheck blockOriginationTime . fmap payloadObj
+
+    -- | Per legal account, the amount of Gas expected to be requested across
+    -- all transactions.
+    --
+    gasTotals :: Balances -> Vector ChainwebTransaction -> GasTotals
+    gasTotals bs = V.foldl' f mempty
+      where
+        f :: GasTotals -> ChainwebTransaction -> GasTotals
+        f gt tx
+            | not (HM.member sender bs) = gt
+            | otherwise = HM.insertWith (+) sender limitInCoin gt
+          where
+            pm = P._pMeta . payloadObj $ P._cmdPayload tx
+            sender = P._pmSender pm
+            P.GasLimit (P.ParsedInteger limit) = P._pmGasLimit pm
+            P.GasPrice (P.ParsedDecimal price) = P._pmGasPrice pm
+            limitInCoin = price * fromIntegral limit
+
+    -- | The balances of all /relevant/ accounts in this group of Transactions.
+    -- TXs which are missing an entry in the `HM.HashMap` should not be
+    -- considered for further processing!
+    --
+    balances :: Vector ChainwebTransaction -> IO Balances
+    balances = foldlM balLookup mempty
+
+    balLookup :: Balances -> ChainwebTransaction -> IO Balances
+    balLookup acc tx
+        | HM.member sender acc = pure acc
+        | otherwise = do
+            let pactHash = view P.cmdHash tx
+            mb <- _cpLookupProcessedTx cp pactHash
```

Comment:
The dupe check shouldn't be moved down into the balance lookup like this, it's really unexpected! Also the semantics are now wrong, if the dupe check fails we need to fail the validation.

Reply:
It effectively does fail it in future steps. Having an entry in the `Balances` map means "we can attempt moving this TX forward."